File: src/api/schemas/knowledge.py
"""Pydantic schemas for knowledge base endpoints (S3)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class Pagination(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int


# ── 漏洞知识库 ────────────────────────────────────────────
class VulnKnowledgeCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=300)
    severity: str = Field(..., pattern="^(critical|high|medium|low|info)$")
    vuln_id: str | None = Field(default=None, max_length=100)
    vuln_type: str | None = None
    vendor: str | None = None
    product: str | None = None
    version_range: str | None = None
    affected_scope: str | None = None
    fingerprint_id: UUID | None = None
    poc: str | None = None
    remediation: str | None = None
    data_source: str | None = None
    upstream_ref: str | None = None


class VulnKnowledgeUpdateRequest(BaseModel):
    name: str | None = None
    severity: str | None = Field(default=None, pattern="^(critical|high|medium|low|info)$")
    vuln_id: str | None = None
    vuln_type: str | None = None
    vendor: str | None = None
    product: str | None = None
    version_range: str | None = None
    affected_scope: str | None = None
    fingerprint_id: UUID | None = None
    poc: str | None = None
    remediation: str | None = None


class ReviewRequest(BaseModel):
    """审核动作: submit(提交审核) / approve(通过) / reject(驳回) / deprecate(弃用)."""
    action: str = Field(..., pattern="^(submit|approve|reject|deprecate)$")
    review_comment: str | None = None


class VulnKnowledgeResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    name: str
    severity: str
    vuln_id: str | None = None
    vuln_type: str | None = None
    vendor: str | None = None
    product: str | None = None
    version_range: str | None = None
    affected_scope: str | None = None
    fingerprint_id: UUID | None = None
    poc: str | None = None
    remediation: str | None = None
    data_source: str | None = None
    upstream_ref: str | None = None
    status: str
    contributed_by: str | None = None
    reviewed_by: str | None = None
    review_comment: str | None = None
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class VulnKnowledgeListResponse(BaseModel):
    data: list[VulnKnowledgeResponse]
    pagination: Pagination


# ── Payloads (合并 字典/关键词) ───────────────────────────
class PayloadCreateRequest(BaseModel):
    name: str | None = None
    content: str = Field(..., min_length=1)
    category: str = Field(..., pattern="^(pass|path|user|header|payload|keywords)$")
    group_name: str | None = None
    weight: float = Field(default=1.0, ge=0)
    description: str | None = None
    data_source: str | None = None


class PayloadUpdateRequest(BaseModel):
    name: str | None = None
    content: str | None = None
    category: str | None = Field(default=None, pattern="^(pass|path|user|header|payload|keywords)$")
    group_name: str | None = None
    weight: float | None = Field(default=None, ge=0)
    description: str | None = None


class PayloadResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    name: str | None = None
    content: str
    category: str
    group_name: str | None = None
    weight: float
    hit_count: int
    description: str | None = None
    data_source: str | None = None
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class PayloadListResponse(BaseModel):
    data: list[PayloadResponse]
    pagination: Pagination


# ── 指纹库 ────────────────────────────────────────────────
class FingerprintCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    category: str | None = None
    component: str | None = None
    version: str | None = None
    match_type: str | None = None
    match_rule: str = Field(..., min_length=1)
    description: str | None = None
    data_source: str | None = None


class FingerprintResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    name: str
    category: str | None = None
    component: str | None = None
    version: str | None = None
    match_type: str | None = None
    match_rule: str
    description: str | None = None
    status: str
    contributed_by: str | None = None
    reviewed_by: str | None = None
    tags: list[str] = Field(default_factory=list)
    vuln_count: int = 0
    created_at: datetime
    updated_at: datetime


class FingerprintListResponse(BaseModel):
    data: list[FingerprintResponse]
    pagination: Pagination


# ── 数据源 ────────────────────────────────────────────────
class DatasourceCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    platform: str = Field(..., min_length=1, max_length=50)
    api_base_url: str | None = None
    config: str | None = None
    description: str | None = None


class DatasourceResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    name: str
    platform: str
    api_base_url: str | None = None
    config: str | None = None
    description: str | None = None
    is_active: int
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class DatasourceListResponse(BaseModel):
    data: list[DatasourceResponse]
    pagination: Pagination


# ── 攻防手册 ──────────────────────────────────────────────
class HandbookCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    category: str | None = None
    content: str = Field(..., min_length=1)
    summary: str | None = None


class HandbookResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    title: str
    category: str | None = None
    content: str
    summary: str | None = None
    status: str
    contributed_by: str | None = None
    reviewed_by: str | None = None
    tags: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class HandbookListResponse(BaseModel):
    data: list[HandbookResponse]
    pagination: Pagination

File: web/src/types/knowledge.ts
// S3 知识库类型 — 对齐后端 routers/knowledge.py

export type KnowledgeStatus = 'draft' | 'pending_review' | 'published' | 'deprecated';

export type PayloadCategory = 'pass' | 'path' | 'user' | 'header' | 'payload' | 'keywords';

export interface Pagination {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

// ── 漏洞知识库 ────────────────────────────────────────────
export interface VulnKnowledge {
  id: string;
  name: string;
  severity: 'critical' | 'high' | 'medium' | 'low' | 'info';
  vuln_id: string | null;
  vuln_type: string | null;
  vendor: string | null;
  product: string | null;
  version_range: string | null;
  affected_scope: string | null;
  fingerprint_id: string | null;
  poc: string | null;
  remediation: string | null;
  data_source: string | null;
  upstream_ref: string | null;
  status: KnowledgeStatus;
  contributed_by: string | null;
  reviewed_by: string | null;
  review_comment: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface VulnKnowledgeListParams {
  page?: number;
  page_size?: number;
  severity?: string;
  status?: string;
  vuln_type?: string;
  q?: string;
}

export interface VulnKnowledgeCreateRequest {
  name: string;
  severity: string;
  vuln_id?: string;
  vuln_type?: string;
  vendor?: string;
  product?: string;
  version_range?: string;
  affected_scope?: string;
  fingerprint_id?: string;
  poc?: string;
  remediation?: string;
  data_source?: string;
  upstream_ref?: string;
}

export interface ReviewRequest {
  action: 'submit' | 'approve' | 'reject' | 'deprecate';
  review_comment?: string;
}

// ── Payloads (字典/关键词合并) ───────────────────────────
export interface Payload {
  id: string;
  name: string | null;
  content: string;
  category: PayloadCategory;
  group_name: string | null;
  weight: number;
  hit_count: number;
  description: string | null;
  data_source: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface PayloadListParams {
  page?: number;
  page_size?: number;
  category?: PayloadCategory;
  group_name?: string;
  q?: string;
}

export interface PayloadCreateRequest {
  name?: string;
  content: string;
  category: PayloadCategory;
  group_name?: string;
  weight?: number;
  description?: string;
  data_source?: string;
}

export interface PayloadUpdateRequest {
  name?: string;
  content?: string;
  category?: PayloadCategory;
  group_name?: string;
  weight?: number;
  description?: string;
}

// ── 指纹库 ────────────────────────────────────────────────
export interface Fingerprint {
  id: string;
  name: string;
  category: string | null;
  component: string | null;
  version: string | null;
  match_type: string | null;
  match_rule: string;
  description: string | null;
  status: KnowledgeStatus;
  contributed_by: string | null;
  reviewed_by: string | null;
  tags: string[];
  vuln_count: number;
  created_at: string;
  updated_at: string;
}

export interface FingerprintCreateRequest {
  name: string;
  category?: string;
  component?: string;
  version?: string;
  match_type?: string;
  match_rule: string;
  description?: string;
  data_source?: string;
}

// ── 数据源 ────────────────────────────────────────────────
export interface Datasource {
  id: string;
  name: string;
  platform: string;
  api_base_url: string | null;
  config: string | null;
  description: string | null;
  is_active: number;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface DatasourceCreateRequest {
  name: string;
  platform: string;
  api_base_url?: string;
  config?: string;
  description?: string;
}

// ── 攻防手册 ──────────────────────────────────────────────
export interface Handbook {
  id: string;
  title: string;
  category: string | null;
  content: string;
  summary: string | null;
  status: KnowledgeStatus;
  contributed_by: string | null;
  reviewed_by: string | null;
  review_comment: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface HandbookCreateRequest {
  title: string;
  category?: string;
  content: string;
  summary?: string;
}

export interface ListResponse<T> {
  data: T[];
  pagination: Pagination;
}
